euphoria-spark/src/main/java/cz/seznam/euphoria/spark/SortTranslator.java

```diff
@@ -0,0 +1,172 @@
+      // pre-reduce age
+      long timestamp = el.getTimestamp();
+
+      return new SparkElement<>(kw._2(), timestamp, el.getElement());
+    });
+  }
+  
+  /**
+   * Extracts {@link Window} from {@link SparkElement} and assigns timestamp 
+   * according to (optional) eventTimeAssigner.
+   * The result composite key tuple consists of [partitionId, window, sortByKey].
+   */
+  private static class CompositeKeyExtractor
+      implements PairFlatMapFunction<SparkElement, Tuple3<Integer, Window, Comparable>, TimestampedElement> {
+
+    private final UnaryFunction<Object, Integer> keyExtractor;
+    private final UnaryFunction<Object, Comparable> sortByFn;
+    private final Windowing windowing;
+    @Nullable
+    private final ExtractEventTime eventTimeAssigner;
+
+    public CompositeKeyExtractor(UnaryFunction<Object, Integer> keyExtractor,
+                                 UnaryFunction<Object, Comparable> sortByFn,
+                                 Windowing windowing,
+                                 @Nullable ExtractEventTime eventTimeAssigner) {
+      this.keyExtractor = keyExtractor;
+      this.sortByFn = sortByFn;
+      this.windowing = windowing;
+      this.eventTimeAssigner = eventTimeAssigner;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Iterator<Tuple2<Tuple3<Integer, Window, Comparable>, TimestampedElement>> call(SparkElement wel) 
+        throws Exception {
+      if (eventTimeAssigner != null) {
+        wel.setTimestamp(eventTimeAssigner.extractTimestamp(wel.getElement()));
+      }
+
+      Iterable<Window> windows = windowing.assignWindowsToElement(wel);
+      List<Tuple2<Tuple3<Integer, Window, Comparable>, TimestampedElement>> out = new ArrayList<>();
+      for (Window wid : windows) {
+        Object el = wel.getElement();
+        long stamp = (wid instanceof TimedWindow)
+                ? ((TimedWindow) wid).maxTimestamp()
+                : wel.getTimestamp();
+        out.add(new Tuple2<>(
+            new Tuple3(keyExtractor.apply(el), wid, sortByFn.apply(el)), 
+            new TimestampedElement(stamp, el)));
+      }
+      return out.iterator();
+    }
+  }
+  
+  /**
+   * Adapter between Euphoria {@link Partitioning} and Spark {@link Partitioner}. Takes already computed
+   * partitionId from the triple.
+   */
+  private static class PartitioningWrapper extends Partitioner {
+
+    private final int numPartitions;
+
+    public PartitioningWrapper(int numPartitions) {
+      this.numPartitions = numPartitions;
+    }
+
+    @Override
+    public int numPartitions() {
+      return numPartitions;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public int getPartition(Object el) {
+      Tuple3<Integer, Window, Comparable> t = (Tuple3<Integer, Window, Comparable>) el;
+      return t._1();
+    }
+  }
+  
+  private static class TripleComparator implements Comparator<Tuple3<Integer, Window, Comparable>>, Serializable {
```

Comment:
i think we should strongly consider switching over to scala. all of this "triple-comparator business" should be this in scala:

```scala
import scala.math.Ordering.Implicits._
var x = (1,2,3) < (1,2,4)
```